core/tests/puppeteer-acceptance-tests/user-utilities/exploration-creator-utils.ts

```diff
@@ -0,0 +1,649 @@
+       * Extract words within parentheses using regex العربية (Arabic) -> Arabic
+       */
+      const matches = selectedLanguageText.match(/\(([^)]+)\)/);
+      if (matches && matches.length >= 2) {
+        return matches[1].trim();
+      } else {
+        // If no parentheses found, assume the whole text is the language.
+        // For eg : English.
+        return selectedLanguageText.trim();
+      }
+    });
+
+    if (selectedLanguage === expectedLanguage) {
+      showMessage(
+        `The language ${selectedLanguage}` + ' is same as expectedLanguage.'
+      );
+    } else {
+      throw new Error('Language is not correct.');
+    }
+    await this.page.keyboard.press('Enter');
+  }
+
+  /**
+   * This function helps in adding tags.
+   */
+  async addTags(TagNames: string[]): Promise<void> {
+    for (let i = 0; i < TagNames.length; i++) {
+      await this.clickOn(addTags);
+      await this.type(addTags, TagNames[i].toLowerCase());
+      await this.page.keyboard.press('Tab');
+    }
+  }
+
+  /**
+   * This function checks if tags are successfully added.
+   */
+  async expectTagsToBeAdded(expectedTags: string[]): Promise<void> {
+    const lowercaseExpectedTags = expectedTags.map(tag => tag.toLowerCase());
+    await this.page.waitForSelector('mat-chip-list');
+
+    const addedTags = await this.page.evaluate(() => {
+      const tagElements = Array.from(document.querySelectorAll('mat-chip'));
+      return tagElements
+        .map(tag => {
+          const textContent =
+            tag.querySelector('.mat-chip-remove')?.previousSibling?.textContent;
+          return textContent ? textContent.trim() : '';
+        })
+        .filter(Boolean);
+    });
+
+    for (const expectedTag of lowercaseExpectedTags) {
+      if (!addedTags.includes(expectedTag)) {
+        throw new Error(`Tag "${expectedTag}" was not added.`);
+      }
+    }
+
+    showMessage('All expected tags were added successfully.');
+  }
+
+  /**
+   * This function allows you to preview the exploration.
+   */
+  async previewSummary(): Promise<void> {
+    await this.page.waitForSelector(
+      '.e2e-test-open-preview-summary-modal:not([disabled])'
+    );
+    await this.clickOn(previewSummaryButton);
+    await this.clickOn(dismissPreviewButton);
+    await this.expectPreviewSummaryToBeVisible();
+  }
+
+  /**
+   * This function verifies that the preview summary is visible.
+   */
+  async expectPreviewSummaryToBeVisible(): Promise<void> {
+    await this.page.waitForSelector('.e2e-test-open-preview-summary-modal');
+    const previewSummary = await this.page.$(
+      '.e2e-test-open-preview-summary-modal'
+    );
+
+    if (previewSummary) {
+      showMessage('Preview summary is visible.');
+    } else {
+      throw new Error('Preview summary is not visible.');
+    }
+  }
+
+  /**
+   * This function helps in updating advanced settings
+   */
+  async enableAutomaticTextToSpeech(): Promise<void> {
+    await this.clickOn(textToSpeechToggle);
```

Comment:
This just clicks on the toggle but doesn't necessarily enable it. To actually verify it's enabled, you'll need to do a check that the post-state of the toggle is correct.

Reply:
We are checking that in next method -> 
```typescript
  async expectAutomaticTextToSpeechToBeEnable(): Promise<void> {
    await this.page.waitForSelector('#text-speech-switch');
    const autoTTSswitch = await this.page.$('#text-speech-switch');
    const isAutoTTSwitchOn = await this.page.evaluate(
      switchElement => switchElement.checked,
      autoTTSswitch
    );
    if (isAutoTTSwitchOn) {
      showMessage('Automatic Text-to-Speech is enabled.');
    } else {
      throw error('Automatic Text-to-Speech is disabled.');
    }
  }
```

Reply:
No, I mean, this method in isolation doesn't guarantee that the toggle gets enabled (and in general, we need to ensure that functions do what they say in their name/docstring). So if some other function uses this method in the future, it might assume that the toggle ends up in the "enabled" position when it actually isn't. So this method itself needs to be hardened.

In general, when fixing a method, that should be done independently of what the method's caller does. That way you can rely on the method to work correctly in all future cases too.